src/apputils/clientsession.ts

```diff
@@ -0,0 +1,819 @@
+  set preferredKernelName(value: string) {
+    this._preferredKernelName = value;
+  }
+
+  /**
+   * The desired kernel language.
+   */
+  get preferredKernelLanguage(): string {
+    return this._preferredKernelLanguage;
+  }
+  set preferredKernelLanguage(value: string) {
+    this._preferredKernelLanguage = value;
+  }
+
+  /**
+   * The type of the client session.
+   */
+  get type(): string {
+    return this._type;
+  }
+
+  /**
+   * Test whether the context is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Dispose of the resources held by the context.
+   */
+  dispose(): void {
+    if (this._isDisposed) {
+      return;
+    }
+    this._isDisposed = true;
+    if (this._session) {
+      this._session.dispose();
+    }
+    Signal.clearData(this);
+  }
+
+  /**
+   * Change the current kernel associated with the document.
+   */
+  changeKernel(options: Kernel.IModel): Promise<Kernel.IKernel> {
+    let session = this._session;
+    if (session) {
+      return session.changeKernel(options);
+    } else if (this._promise) {
+      return this._promise.promise.then(() => {
+        return session.changeKernel(options);
+      });
+    } else {
+      return this._startSession(options);
+    }
+  }
+
+  /**
+   * Kill the kernel and shutdown the session.
+   *
+   * @returns A promise that resolves when the session is shut down.
+   */
+  shutdown(): Promise<void> {
+    if (this._session) {
+      return this._session.shutdown();
+    } else if (this._promise) {
+      return this._promise.promise.then(() => {
+        return this._session.shutdown();
+      });
+    }
+    return Promise.resolve(void 0);
+  }
+
+  /**
+   * Restart the session.
+   *
+   * @returns A promise that resolves with the kernel model.
+   *
+   * #### Notes
+   * If there is a running kernel, present a dialog.
+   * If there is no kernel, we start a kernel with the last run
+   * kernel name.  If no kernel has been started, this is a no-op.
+   */
+  restart(): Promise<Kernel.IKernel | null> {
+    let kernel = this.kernel;
+    if (!kernel) {
+      if (this._prevKernelName) {
+        return this.changeKernel({ name: this._prevKernelName });
+      }
+      return Promise.resolve(null);
+    }
+    return ClientSession.restartKernel(kernel);
+  }
+
+  /**
+   * Select a kernel for the session.
+   */
+  selectKernel(): Promise<void> {
+    return Private.selectKernel(this, this._manager);
+  }
+
+  /**
+   * Change the session path.
+   *
+   * @param path - The new session path.
+   *
+   * @returns A promise that resolves when the session has renamed.
+   *
+   * #### Notes
+   * This uses the Jupyter REST API, and the response is validated.
+   * The promise is fulfilled on a valid response and rejected otherwise.
+   */
+  setPath(path: string): Promise<void> {
+    this._path = path;
+    if (this._session) {
+      return this._session.rename(path);
+    }
+    return Promise.resolve(void 0);
```

Comment:
No, we want to set the path regardless of whether there is a session on the server.